### ISSUE TITLE
Avoid that cases with many contacts make the persons query conditions to grow exponentially [3]

### ISSUE THREAD

**Issue by StefanKock** [hzi-braunschweig/SORMAS-Project#6188]
### Problem Description
Follow-up to #6170: Having some cases with many contacts each causes the user filter conditions to explode for the persons directory.

<details>
<summary>Problem analysis</summary>

The problem already starts with only case persons and contact persons in the persons directory, having event participants in there increases the problem: https://explain.dalibo.com/plan/71C

I think I found the reason for the data explosion: some cases have hundreds of contacts.
```sql
SELECT caze_id, count(*) FROM contact GROUP BY caze_id ORDER BY COUNT DESC;

 caze_id | count
---------+-------
     314 |   544
 2270665 |   224
         |    97
     454 |    27
    1588 |    13
 2282909 |     8
     145 |     5
 2283801 |     5
 1830515 |     4
```

</details>

### Proposed Change

Detailing option 2 from below.

1. Make `PersonCriteria.getPersonAssociation` mandatory (`null` equals `PersonAssociation.ALL`) because `null` is not a useful option.
- [x] Alter `PersonCriteria` that `personAssociation` is always set, defaulting to `ALL`.
- [x] Remove superfluous `null` checks in the backend.

2. Improve `PersonFacade.count`, `PersonFacade.getIndexList`: If only one association is queried, only ask for that related persons.
- [x] Remove superfluous subselects in `PersonService.buildCriteriaFilter` (it's already done by `PersonService.createUserFilter` since #6170).
- [x] Switch the association join to Inner Join.

3. Speed up `PersonFacade.count` with `PersonAssociation.ALL`.
- [x] Run ~in parallel~ queries per `PersonAssociation` to fetch person ids, put all ids into a `Set` to find out the count.

4. Speed up `PersonFacade.getIndexList` with `PersonAssociation.ALL`.
- [ ] Not working idea, out of scope: Run ~in parallel~ queries per `PersonAssociation` to fetch person ~ids~ uuids **with** sorting and offset & limit, put all ids into a `Set`, run a second query with filtering by person ~ids~ uuids and repeated sorting and limit. (changed to uuids: For sorting all sortable attributes need to be present in the result, so the same getIndexList DTO multiselect is used, and there is the uuid instead of the id available).

5. Enhance logging.
- [x] Debug log for duration of `PersonFacade.count` (log `personAssciation` and `count`) and `PersonFacade.getIndexList` (log `personAssciation` and list size).

6. Check that all use cases aside still work.
- [x] `PersonFacadeEjb.getAllUuidsBatched` -> `PersonFacadeEjb..setMissingGeoCoordinates`

### Possible Alternatives

<details>
<summary>1. Reducing the JOIN complexity here by fetching the person ids per entity of interest and then UNION the result in SQL.</summary>

Pseudo code for possible new query structure:
```sql
SELECT DISTINCT attributes FROM person

-- For plain attributes
LEFT OUTER JOIN location
LEFT OUTER JOIN district

-- For calculated attribute isInJurisdiction
LEFT OUTER JOIN case
LEFT OUTER JOIN contact
LEFT OUTER JOIN eventparticipant
LEFT OUTER JOIN location
LEFT OUTER JOIN event
LEFT OUTER JOIN immunization
LEFT OUTER JOIN travelentry

WHERE person.id IN
(
SELECT person.id FROM person -- JOIN WHERE case userFilter
UNION
SELECT person.id FROM person -- JOIN WHERE contact userFilter
UNION
SELECT person.id FROM person -- JOIN WHERE eventparticipant userFilter
UNION
SELECT person.id FROM person -- JOIN WHERE immunization userFilter
UNION
SELECT person.id FROM person -- JOIN WHERE travelentry userFilter
)
```

Problem: My research showed that JPA and Hibernate don't support UNIONs. It would be possible with native SQL, but all current queries are dynamically built with Criteria API.

I ran an analysis for the `SELECT person.id FROM person -- JOIN WHERE contact userFilter` part: https://explain.dalibo.com/plan/SVo

</details>

<details>
<summary>2. Do the option 1 in two steps via JPA</summary>

1. Find the person Ids with a query per association (with offset and limit with adhered order). Maybe this can be done in parallel by Java code.
2. Union the results in Java by fetching all found person ids.
3. Query the data into DTOs (order, distinct, limit).

Expected limitation: Does only work for lazy loading, around above 100 entries. There should be a limit to about 200 persons to be fetched at once.
Result: It should work almost as fast as the option 1, and all current filtering and query builder code can be reused.
Sidenote: The Queries from step 1 also need to be run for the count Query without filtering and limit.

</details>

<details>
<summary>3. Redundantly precompile to what jurisdictions the associated entities belong</summary>

Unclear how, can be quite large. We would need to cover all variations that are currently in the code.

</details>

### Additional Information

Ideas for another ticket:
- Maybe make the persons view not directly fetch data but let the user select what he needs.
- Force the user to choose at least some filtering in the persons view?

**Comment by StefanKock** [comment]
Waiting for refinement meeting tomorrow

**Comment by StefanKock** [comment]
I did a benchmark with the worst execution example I know from previous tickets. Summary: I pushed down the time to less than 1s (instead of minutes to over and hour).

- [#6188 PersonFacade associations New Queries.sql.txt](https://github.com/hzi-braunschweig/SORMAS-Project/files/6899948/6188.PersonFacade.associations.New.Queries.sql.txt)
- [#6188 PersonFacade associations Benchmark.xlsx](https://github.com/hzi-braunschweig/SORMAS-Project/files/6899949/6188.PersonFacade.associations.Benchmark.xlsx) (with links to analysis per query)

**Comment by StefanKock** [comment]
I have a problem with changes of `PersonFacadeEjb.getIndexList`: The unit test `PersonFacadeEjbUserFilterTest.testGetPersonIndexListWhenSeveralAssociations` now fails, because one person is listed twice (one pseudonymized, one not).

<details>
<summary>Query generates duplicate entity</summary>

```

java.lang.AssertionError: expected:<2> but was:<3>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at org.junit.Assert.assertEquals(Assert.java:633)
	at de.symeda.sormas.backend.person.PersonFacadeEjbUserFilterTest.testGetPersonIndexListWhenSeveralAssociations(PersonFacadeEjbUserFilterTest.java:126)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
...

```

```sql
SELECT DISTINCT person0_.uuid AS col_0_0_,
                person0_.firstName AS col_1_0_,
                person0_.lastName AS col_2_0_,
                person0_.approximateAge AS col_3_0_,
                person0_.approximateAgeType AS col_4_0_,
                person0_.birthdate_dd AS col_5_0_,
                person0_.birthdate_mm AS col_6_0_,
                person0_.birthdate_yyyy AS col_7_0_,
                person0_.sex AS col_8_0_,
                district2_.name AS col_9_0_,
                location1_.street AS col_10_0_,
                location1_.houseNumber AS col_11_0_,
                location1_.postalCode AS col_12_0_,
                location1_.city AS col_13_0_,

  (SELECT personcont12_.contactInformation
   FROM PersonContactDetail personcont12_
   WHERE personcont12_.person_id=person0_.id
     AND personcont12_.primaryContact=1
     AND personcont12_.personContactDetailType='PHONE') AS col_14_0_,

  (SELECT personcont13_.contactInformation
   FROM PersonContactDetail personcont13_
   WHERE personcont13_.person_id=person0_.id
     AND personcont13_.primaryContact=1
     AND personcont13_.personContactDetailType='EMAIL') AS col_15_0_,
                person0_.changeDate AS col_16_0_,
                CASE
                    WHEN 0=1
                         OR (cases3_.reportingUser_id IS NOT NULL)
                         AND cases3_.reportingUser_id=17
                         OR cases3_.responsibleDistrict_id=9
                         OR cases3_.district_id=9
                         OR (contacts4_.reportingUser_id IS NOT NULL)
                         AND contacts4_.reportingUser_id=17
                         OR contacts4_.district_id=9
                         OR case5_.responsibleDistrict_id=9
                         OR case5_.district_id=9
                         OR (eventparti6_.reportingUser_id IS NOT NULL)
                         AND eventparti6_.reportingUser_id=17
                         OR eventparti6_.district_id=9
                         OR location8_.district_id=9
                         OR (travelentr9_.reportingUser_id IS NOT NULL)
                         AND travelentr9_.reportingUser_id=17
                         OR travelentr9_.responsibleDistrict_id=9
                         OR travelentr9_.pointOfEntryDistrict_id=9
                         OR (immunizati11_.reportingUser_id IS NOT NULL)
                         AND immunizati11_.reportingUser_id=17
                         OR immunizati11_.responsibleDistrict_id=9 THEN 1
                    ELSE 0
                END AS col_17_0_
FROM Person person0_
LEFT OUTER JOIN LOCATION location1_ ON person0_.address_id=location1_.id
LEFT OUTER JOIN District district2_ ON location1_.district_id=district2_.id
LEFT OUTER JOIN cases cases3_ ON person0_.id=cases3_.person_id
LEFT OUTER JOIN Contact contacts4_ ON person0_.id=contacts4_.person_id
LEFT OUTER JOIN cases case5_ ON contacts4_.caze_id=case5_.id
LEFT OUTER JOIN EventParticipant eventparti6_ ON person0_.id=eventparti6_.person_id
LEFT OUTER JOIN EVENTS event7_ ON eventparti6_.event_id=event7_.id
LEFT OUTER JOIN LOCATION location8_ ON event7_.eventLocation_id=location8_.id
LEFT OUTER JOIN travelentry travelentr9_ ON person0_.id=travelentr9_.person_id
LEFT OUTER JOIN cases case10_ ON travelentr9_.resultingCase_id=case10_.id
LEFT OUTER JOIN immunization immunizati11_ ON person0_.id=immunizati11_.person_id
WHERE person0_.uuid in ('UM3WR4-M6PYBV-MWLBEQ-NROLSPFM',
                        'XCVOWR-2RI3VH-M2YBOE-7J3RCEMA')
ORDER BY person0_.changeDate DESC
```

</details>

**Comment by StefanKock** [comment]
> I have a problem with changes of `PersonFacadeEjb.getIndexList`: The unit test `PersonFacadeEjbUserFilterTest.testGetPersonIndexListWhenSeveralAssociations` now fails, because one person is listed twice (one pseudonymized, one not).

I've an idea to solve the Problem: Not selecting the jurisdiction will the other attributes in the query, but selecting it in a secondary query with the same filter (but not paging and sorting), then updating the boolean attribute on the `List<PersonIndexDto>` entries. Overall the impact should be that bad, because the first query looses a lot of complexity, especially for `PersonAssociation.ALL`.

**Comment by StefanKock** [comment]
Ticket is now fully implemented. Deviations from description:
1. No parallel queries, would need more sophisticated structure concerning EJBs and JTA transactions than just `parallelStream`.
2. Using `getIndexList` query to fetch `personUuids`: Full attribute query needed for correct sorting and paging, and since it does not contain the `Person.id`, I now use uuids to do the same.
3. I solved the problem described in https://github.com/hzi-braunschweig/SORMAS-Project/issues/6188#issuecomment-892682695 by splitting out the jurisdiction query as a secondary query.
4. Development started on `1.63.0-SNAPSHOT` (at start of version) and implemented all commits, then cherry-picked all commits into another branch based on `hotfix-1.62.2`. I successfully ran all backend unit tests on both branches.
